barter-data/src/exchange/coinbase/book/l2.rs

```diff
@@ -0,0 +1,665 @@
+            let market = sub.id();
+            let snapshot_url = format!(
+                "{url}/products/{product_id}/book?level=2",
+                url = HTTP_COINBASE_API,
+                product_id = market.as_ref(),
+            );
+
+            async move {
+                let client = Client::new();
+
+                // Fetch initial OrderBook snapshot via HTTP
+                let snapshot = client
+                    .get(&snapshot_url)
+                    .header("User-Agent", "barter-rs")
+                    .send()
+                    .await
+                    .map_err(SocketError::Http)?
+                    .json::<CoinbaseOrderBookSnapshot>()
+                    .await
+                    .map_err(SocketError::Http)?;
+
+                Ok(MarketEvent::from((
+                    ExchangeId::Coinbase,
+                    sub.instrument.key().clone(),
+                    snapshot,
+                )))
+            }
+        });
+
+        try_join_all(l2_snapshot_futures)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct CoinbaseOrderBookSnapshotLevel {
+    #[serde(with = "rust_decimal::serde::str")]
+    pub price: Decimal,
+    #[serde(with = "rust_decimal::serde::str")]
+    pub size: Decimal,
+    pub num_orders: u32,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct CoinbaseOrderBookL2SnapshotLevel {
+    #[serde(with = "rust_decimal::serde::str")]
+    pub price: Decimal,
+    #[serde(with = "rust_decimal::serde::str")]
+    pub size: Decimal
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct CoinbaseOrderBookL2Update {
+    pub product_id: String,
+    #[serde(
+        alias = "time"
+    )]
+    pub time_exchange: DateTime<Utc>,
+    pub changes: Vec<CoinbaseOrderBookL2Change>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+#[serde(untagged)]
+pub enum CoinbaseOrderBookL2Message {
+    Snapshot(CoinbaseOrderBookL2Snapshot),
+    Update(CoinbaseOrderBookL2Update),
+}
+
+impl Identifier<Option<SubscriptionId>> for CoinbaseOrderBookL2Message {
+    fn id(&self) -> Option<SubscriptionId> {
+        match self {
+            CoinbaseOrderBookL2Message::Snapshot(snapshot) => Some(SubscriptionId::from(format!("level2_batch|{}",  snapshot.product_id.clone()))),
+            CoinbaseOrderBookL2Message::Update(update) => Some(SubscriptionId::from(format!("level2_batch|{}",  update.product_id.clone()))),
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct CoinbaseOrderBookL2Change {
+    pub side: Side,
+    #[serde(with = "rust_decimal::serde::str")]
+    pub price: Decimal,
+    #[serde(with = "rust_decimal::serde::str")]
+    pub size: Decimal,
+}
+
+#[derive(Debug, Constructor)]
+pub struct CoinbaseOrderBookL2Meta<InstrumentKey, Sequencer> {
+    pub key: InstrumentKey,
+    pub sequencer: Sequencer,
+}
+
+#[derive(Debug, Default)]
+pub struct CoinbaseOrderBookL2Sequencer {
```

Comment:
I noticed that (one of) the OrderBook structs contains a `sequence` - should this not be utilised in the sequencer?